src/program/mod.rs

```diff
@@ -395,35 +474,14 @@ impl<'a> Program<'a> {
 
 /// make dst determines the parameters of a rectangle required to place an image correctly in
 /// the window
-fn make_dst(src_x: u32, src_y: u32, dst_x: u32, dst_y: u32) -> Rect {
-    // case 1: both source dimensions smaller
-    if src_x < dst_x && src_y < dst_y {
-        return full_rect(src_x, src_y, dst_x, dst_y);
-    }
-    // case 2: source aspect ratio is larger
-    if src_x as f32 / src_y as f32 > dst_x as f32 / dst_y as f32 {
-        return fit_x_rect(src_x, src_y, dst_x, dst_y);
-    }
-    // case 3: source aspect ratio is smaller
-    fit_y_rect(src_x, src_y, dst_x, dst_y)
-}
-
-fn full_rect(src_x: u32, src_y: u32, dst_x: u32, dst_y: u32) -> Rect {
-    let y = ((dst_y - src_y) as f32 / 2.0) as i32;
-    let x = ((dst_x - src_x) as f32 / 2.0) as i32;
-    Rect::new(x, y, src_x, src_y)
-}
-
-fn fit_x_rect(src_x: u32, src_y: u32, dst_x: u32, dst_y: u32) -> Rect {
-    let height = ((src_y as f32 / src_x as f32) * dst_x as f32) as u32;
-    let y = ((dst_y - height) as f32 / 2.0) as i32;
-    Rect::new(0, y, dst_x, height)
-}
-
-fn fit_y_rect(src_x: u32, src_y: u32, dst_x: u32, dst_y: u32) -> Rect {
-    let width = ((src_x as f32 / src_y as f32) * dst_y as f32) as u32;
-    let x = ((dst_x - width) as f32 / 2.0) as i32;
-    Rect::new(x, 0, width, dst_y)
+fn make_dst(tq: &TextureQuery, vp: &Rect, scale: f32, pan_x: f32, pan_y: f32) -> Rect {
+    let x_diff = (vp.width() as f32 - (tq.width as f32 * scale)) / 2.0;
+    let y_diff = (vp.height() as f32 - (tq.height as f32 * scale)) / 2.0;
+    let x = (x_diff - x_diff * pan_x) as i32;
+    let y = (y_diff - y_diff * pan_y) as i32;
```

Comment:
A problem I've had with panning amount being based off image dimensions is that for really large images the pan amount is almost a full page. Zooming in makes the pan amount seem even bigger.

Maybe it could be based off a combination of window size and zoom level?
Example large image
[screencapture-hoverbear-org-2016-10-12-rust-state-machine-pattern-2019-05-25-19_57_08](https://user-images.githubusercontent.com/28968644/58376116-4a17bd80-7f28-11e9-9eeb-cf6c8d0a5334.png)

Reply:
👍